libraries/SITL/examples/JSON/C++/libAP_JSON.cpp

```diff
@@ -0,0 +1,263 @@
+/*
+ * This file is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This file is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+/*
+Adapted from Rhys Mainwaring's ArduPilot plugin for Gazebo
+*/
+
+#include "libAP_JSON.h"
+
+#define DEBUG_ENABLED 1
+
+// SITL JSON interface supplies 16 servo channels
+#define MAX_SERVO_CHANNELS 16
+
+// The servo packet received from ArduPilot SITL. Defined in SIM_JSON.h.
+struct servo_packet
+{
+    uint16_t magic; // 18458 expected magic value
+    uint16_t frame_rate;
+    uint32_t frame_count;
+    uint16_t pwm[16];
+};
+
+bool libAP_JSON::InitSockets(const char *fdm_address, const uint16_t fdm_port_in)
+{
+    // configure port
+    sock.set_blocking(false);
+    sock.reuseaddress();
+
+    // bind the socket
+    if (!sock.bind(fdm_address, fdm_port_in))
+    {
+        std::cout << "[libAP_JSON] "
+                  << "failed to bind with "
+                  << fdm_address << ":" << fdm_port_in
+                  << std::endl;
+        return false;
+    }
+    std::cout << "[libAP_JSON] " << "flight dynamics model at "
+              << fdm_address << ":" << fdm_port_in
+              << std::endl;
+    return true;
+}
+
+bool libAP_JSON::ReceiveServoPacket(uint16_t servo_out[])
+{
+    // Added detection for whether ArduPilot is online or not.
+    // If ArduPilot is detected (receive of fdm packet from someone),
+    // then socket receive wait time is increased from 1ms to 1 sec
+    // to accommodate network jitter.
+    // If ArduPilot is not detected, receive call blocks for 1ms
+    // on each call.
+    // Once ArduPilot presence is detected, it takes this many
+    // missed receives before declaring the FCS offline.
+
+    uint32_t waitMs;
+    if (arduPilotOnline)
+    {
```

Comment:
```suggestion
    if (arduPilotOnline) {
```


... and generally ardupilot-ise

Reply:
Fixed up the ones I recognize.